### PR TITLE
Firefox: Pass title tests

### PR DESCRIPTION
**Pull request by Meir017**
we don't have an issue just for page-title tests.
this adds the evaluate feature (still missing some evaluate features)

Relates to #113

### CODE REVIEW

File: src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs [hunk 1]
```diff
@@ -13,14 +18,166 @@ public FirefoxExecutionContext(FirefoxSession workerSession, string executionCon
...
+            string functionText = pageFunction.ToFunctionText();
+            var callFunctionTask = _session.SendAsync(new RuntimeCallFunctionRequest
+            {
+                FunctionDeclaration = functionText,
+                Args = Array.ConvertAll(args, FormatArgument),
+                ReturnByValue = returnByValue,
+                ExecutionContextId = _executionContextId,
+            });
+
+            // TODO: validate request
```
Comment (Meir017) [inline src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs]: this will be added for tests that use the arguments feature

**Comment by kblok** [comment]
@Meir017 this PR would be part of #113 

Issues are created based on PW tests

File: src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs [hunk 2]
```diff
@@ -13,14 +18,166 @@ public FirefoxExecutionContext(FirefoxSession workerSession, string executionCon
...
+            {
+                var payload = await callFunctionTask.ConfigureAwait(false);
+                return ExtractResult(payload.ExceptionDetails, payload.Result);
+            }
+
+            T ExtractResult(ExceptionDetails exceptionDetails, RemoteObject remoteObject)
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs]: Can we move these methods outside `EvaluateAsync`?

File: src/PlaywrightSharp/Helpers/StringExtensions.cs [hunk 3]
```diff
@@ -91,6 +91,16 @@ public static bool IsJavascriptFunction(this string script)
             return false;
         }
 
+        /// <summary>
+        /// Converts the page function to a function text.
+        /// </summary>
+        /// <param name="pageFunction">The page function.</param>
+        /// <returns>The valid function text.</returns>
+        public static string ToFunctionText(this string pageFunction)
```
Comment (kblok) [inline src/PlaywrightSharp/Helpers/StringExtensions.cs]: We don't need this.

Comment (Meir017) [reply src/PlaywrightSharp/Helpers/StringExtensions.cs]: OK, I wasn't sure about this

File: src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs [hunk 4]
```diff
@@ -13,14 +18,166 @@ public FirefoxExecutionContext(FirefoxSession workerSession, string executionCon
...
+
+        private CallFunctionArgument FormatArgument(object arg, FrameExecutionContext context)
+        {
+            switch (arg)
+            {
+                case int integer when integer == -0:
+                    return new CallFunctionArgument { UnserializableValue = RemoteObjectUnserializableValue.NegativeZero };
+
+                case double d:
+                    if (double.IsPositiveInfinity(d))
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs]: Can't we use `when`s where?

File: src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs [hunk 5]
```diff
@@ -13,14 +18,166 @@ public FirefoxExecutionContext(FirefoxSession workerSession, string executionCon
...
+            if (!pageFunction.IsJavascriptFunction())
+            {
+                var payload = await _session.SendAsync(new RuntimeEvaluateRequest
+                {
+                    Expression = pageFunction.Trim(),
+                    ReturnByValue = returnByValue,
+                    ExecutionContextId = _executionContextId,
+                }).ConfigureAwait(false);
+
+                // TODO: rewriteError
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs]: We won't need to rewriteErrors

Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs]: You are already rewriting the error on the ExtractResult method.

File: src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs [hunk 6]
```diff
@@ -13,14 +18,166 @@ public FirefoxExecutionContext(FirefoxSession workerSession, string executionCon
...
+                    return ToCallArgument(objectHandle.RemoteObject);
+            }
+
+            return new CallFunctionArgument
+            {
+                Value = arg,
+            };
+        }
+
+        private CallFunctionArgument ToCallArgument(IRemoteObject remoteObject) => new CallFunctionArgument
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs]: I would put the arrow on the next line

File: src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs [hunk 7]
```diff
@@ -13,14 +18,166 @@ public FirefoxExecutionContext(FirefoxSession workerSession, string executionCon
...
+            }
+
+            return typeof(T) == typeof(JsonElement) ? (T)remoteObject.Value : (T)ValueFromType<T>((JsonElement)remoteObject.Value, remoteObject.Type ?? RemoteObjectType.Object);
+        }
+
+        private object CreateHandle(RemoteObject remoteObject, FrameExecutionContext context) => new JSHandle(context, remoteObject);
+
+        private object ValueFromUnserializableValue(RemoteObject remoteObject, RemoteObjectUnserializableValue unserializableValue)
+        {
+            return unserializableValue switch
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs]: Arrow?

File: src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs [hunk 8]
```diff
@@ -13,14 +18,166 @@ public FirefoxExecutionContext(FirefoxSession workerSession, string executionCon
...
+                RemoteObjectUnserializableValue.NaN => double.NaN,
+                RemoteObjectUnserializableValue.Infinity => double.PositiveInfinity,
+                RemoteObjectUnserializableValue.NegativeInfinity => double.NegativeInfinity,
+                _ => throw new Exception("Unsupported unserializable value: " + unserializableValue),
+            };
+        }
+
+        private object ValueFromType<T>(JsonElement value, RemoteObjectType objectType)
+        {
+            switch (objectType)
```
Comment (kblok) [inline src/PlaywrightSharp.Firefox/FirefoxExecutionContext.cs]: Can't arrow and fancy switch here?

**Pull request merged**